### PR TITLE
FIX: Tests now pass in Postgres on 3.1 and 3.2

### PR DESCRIPTION
**Pull request by gordonbanderson**
Problems resolved:
1) Case sensitivy of check on Code field of Group
2) MySQL and Postgres have different date functions
3) When BlogID is empty, explicitly set it to 0.  If not then all tests break under Postgres

Fix for https://github.com/silverstripe/silverstripe-blog/issues/361

### CODE REVIEW

**Comment by dhensby** [comment]
this needs rebasing :(

if you run `git pull --rebase https://github.com/silverstripe/silverstripe-blog master` then `git push -f` it should work

**Comment by gordonbanderson** [comment]
Rebase done as requested

**Comment by dhensby** [comment]
thanks :)

File: .travis.yml [hunk 1]
```diff
@@ -43,6 +48,8 @@ before_script:
   - git clone git://github.com/silverstripe-labs/silverstripe-travis-support.git ~/travis-support
   - php ~/travis-support/travis_setup.php --source `pwd` --target ~/builds/ss
   - cd ~/builds/ss
+  - cat ./framework/model/DB.php
```
Comment (dhensby) [inline .travis.yml]: do we need these lines?

Comment (gordonbanderson) [inline .travis.yml]: No, that is debug I thought I'd rebased out.  Fix, rebase and git push -f again...

**Comment by gordonbanderson** [comment]
Late my time so if any further issues will have to wait until Monday realistically.

File: code/extensions/URLSegmentExtension.php [hunk 2]
```diff
@@ -40,6 +42,12 @@ public function generateURLSegment($increment = null)
...
 
+        // Postgres use '' instead of 0 as an emtpy blog ID
+        // Without this all the tests fail
+        if (!$this->owner->BlogID) {
+            $this->owner->BlogID = 0;
+        }
+
         $duplicate = DataList::create($this->owner->ClassName)->filter(array(
             'URLSegment' => $this->owner->URLSegment,
             'BlogID' => $this->owner->BlogID,
```
Comment (tractorcow) [inline code/extensions/URLSegmentExtension.php]: If we know BlogID is 0 why are we filtering on it? It won't match anything. The second condition should have been inside an `if()` similar to how the `$this->owner->ID` check is.

In any case, the code path won't really reach here if BlogID isn't set, so it shouldn't really matter in any case.

File: code/model/Blog.php [hunk 3]
```diff
@@ -473,16 +473,38 @@ public function getArchivedBlogPosts($year, $month = null, $day = null)
...
+            if ($month) {
+                $where .= sprintf(' AND EXTRACT(MONTH FROM "PublishDate") = \'%s\'', Convert::raw2sql($month));
+
+                if ($day) {
+                    $where .= sprintf(' AND EXTRACT(DAY FROM "PublishDate") = \'%s\'', Convert::raw2sql($day));
+                }
+            }
+
+            $query->where($where);
         }
```
Comment (tractorcow) [inline code/model/Blog.php]: Else, it just doesn't work? So now blog only works on postgres and mysql?

@dhensby I'm disappointed in this getting merged. =/

@gordonbanderson this is a good try, I'm not blaming you. This should have been given some more review prior to merge, that's all.

**Comment by tractorcow** [comment]
@gordonbanderson thanks for your hard work to try to diagnose this issue. I'll suggest a change to this fix later on today.

**Comment by tractorcow** [comment]
`DB::getConn()->formattedDatetimeClause` is the best way to delegate date formatting syntax to the current connector. Maybe we can use that instead?

**Comment by tractorcow** [comment]
I've thrown up a possible fix at https://github.com/silverstripe/silverstripe-blog/pull/363; I'm just going to let travis-ci take care of all the cross-adaptor testing for me though. :)

File: code/extensions/URLSegmentExtension.php (hunk 2 above)
Comment (gordonbanderson) [inline code/extensions/URLSegmentExtension.php]: Take those 3 lines out and all of the Postgres tests fail.  The issue does not occur with MySQL and SQLite.

File: code/model/Blog.php (hunk 3 above)
Comment (gordonbanderson) [inline code/model/Blog.php]: Previously it only worked with MySQL... re-testing I forked prior to my recent changes and added Travis checks (which were missing) for PGSQL and SQLite, see https://github.com/gordonbanderson/silverstripe-blog/commits/master_20160119_check_dbs

The Travis build fails for both Postgres and SQLite, https://travis-ci.org/gordonbanderson/silverstripe-blog/builds/104570379

**Comment by gordonbanderson** [comment]
I was previously unaware of the `DB::getConn()->formattedDatetimeClause` method

**Comment by dhensby** [comment]
derp :/ sorry - I was obviously a bit too keen when merging this and thought green ticks meant all the DBs were ok. That if/else was quite obvious now looking back.

note to self: must try harder

**Comment by tractorcow** [comment]
Sorry, I didn't mean to come off sounding harsh. :) The process works... peer review saved the day. :P

**Pull request merged**